lib/app.js

```diff
@@ -0,0 +1,69 @@
+const HOOK_STAGE_PRE_INSTALL = 'preInstall'
+const HOOK_STAGE_POST_INSTALL = 'postInstall'
+
+const isObject = o => typeof o === 'object' && o !== null && !Array.isArray(o)
+
+const runHandler = handler => {
+  handler = typeof handler === 'function' ? handler : () => {}
+
+  let resultMaybePromise
+  try {
+    resultMaybePromise = handler()
+  } catch (err) {
+    resultMaybePromise = false
+  }
+
+  const isPromise = resultMaybePromise && typeof resultMaybePromise.then === 'function'
+  const resultPromise = isPromise ? resultMaybePromise : Promise.resolve(resultMaybePromise)
+
+  return resultPromise.then(
+    result => {
+      if (result instanceof Error || result === false) {
+        return false
+      } else if (!isObject(result)) {
+        return {}
+      } else {
+        return result
+      }
+    },
+    () => false
+  )
+}
+
+module.exports = function createApp(channel) {
+  const handlers = {
+    [HOOK_STAGE_PRE_INSTALL]: null,
+    [HOOK_STAGE_POST_INSTALL]: null
+  }
+
+  const setHandler = (stage, handler) => {
+    if (handlers[stage]) {
+      throw new Error('Cannot register a handler twice.')
+    } else if (typeof handler !== 'function') {
+      throw new Error('Handler must be a function.')
```

Comment:
It looks like you `throw` here if the handler isn't a function but you still check for it in the `runHandler` function but don't throw there. So that code path *seems* like it will never run anyway with this code in the way here.

Reply:
This method is used by a user to inject a handler.

This condition protects us against `app.onConfigure('boom')` calls from end-user.

`runHandler` checks if a handler is registered and provides a default if not b/c as per spec it's not required (no handler registered means no parameters will be used for an app which sometimes may be valid).